storage/pruning/pruningStorer.go

```diff
@@ -0,0 +1,444 @@
+				v, err = ps.activePersisters[idx].persister.Get(key)
+
+				if err != nil {
+					continue
+				}
+
+				found = true
+				// if found in persistence unit, add it in cache
+				ps.cacher.Put(key, v)
+				break
+			}
+		}
+		if !found {
+			return nil, fmt.Errorf("key %s not found in %s",
+				base64.StdEncoding.EncodeToString(key), ps.identifier)
+		}
+	}
+
+	return v.([]byte), nil
+}
+
+// GetFromEpoch will search a key only in the persister for the given epoch
+func (ps *PruningStorer) GetFromEpoch(key []byte, epoch uint32) ([]byte, error) {
+	// TODO: this will be used when requesting from resolvers
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	v, ok := ps.cacher.Get(key)
+	if ok {
+		return v.([]byte), nil
+	}
+
+	persisterData, exists := ps.persistersMapByEpoch[epoch]
+	if !exists {
+		return nil, fmt.Errorf("key %s not found in %s",
+			base64.StdEncoding.EncodeToString(key), ps.identifier)
+	}
+
+	if !persisterData.isClosed {
+		return persisterData.persister.Get(key)
+	} else {
+		persister, err := ps.persisterFactory.Create(persisterData.path)
+		if err != nil {
+			log.Debug("open old persister", "error", err.Error())
+			return nil, err
+		}
+
+		err = persister.Init()
+		if err != nil {
+			log.Debug("init old persister", "error", err.Error())
+			return nil, err
+		}
+
+		res, errToRet := persister.Get(key)
+
+		err = persister.Close()
+		if err != nil {
+			log.Debug("close old persister", "error", err.Error())
+			return nil, err
+		}
+
+		if errToRet == nil {
+			return res, nil
+		}
+		log.Warn("get from closed persister",
+			"id", ps.identifier,
+			"epoch", epoch,
+			"key", key,
+			"error", errToRet.Error())
+
+		return nil, fmt.Errorf("key %s not found in %s",
+			base64.StdEncoding.EncodeToString(key), ps.identifier)
+	}
+}
+
+// Has checks if the key is in the Unit.
+// It first checks the cache. If it is not found, it checks the bloom filter
+// and if present it checks the db
+func (ps *PruningStorer) Has(key []byte) error {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	has := ps.cacher.Has(key)
+	if has {
+		return nil
+	}
+
+	if ps.bloomFilter == nil || ps.bloomFilter.MayContain(key) {
+		for _, persister := range ps.activePersisters {
+			if persister.persister.Has(key) != nil {
+				continue
+			}
+
+			return nil
+		}
+	}
+
+	return storage.ErrKeyNotFound
+}
+
+// HasInEpoch checks if the key is in the Unit in a given epoch.
+// It first checks the cache. If it is not found, it checks the bloom filter
+// and if present it checks the db
+func (ps *PruningStorer) HasInEpoch(key []byte, epoch uint32) error {
+	// TODO: this will be used when requesting from resolvers
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	has := ps.cacher.Has(key)
+	if has {
+		return nil
+	}
+
+	if ps.bloomFilter == nil || ps.bloomFilter.MayContain(key) {
+		persister, ok := ps.persistersMapByEpoch[epoch]
+		if !ok {
+			return storage.ErrKeyNotFound
+		}
+
+		return persister.persister.Has(key)
```

Comment:
how do you know the persister is opened?

Reply:
very good catch. implemented the same logic from GetFromEpoch now